receiver/aerospikereceiver/scraper.go

```diff
@@ -0,0 +1,217 @@
+	config   *Config
+	consumer consumer.Metrics
+	host     string // host/IP of configured Aerospike node
+	port     int    // port of configured Aerospike node
+	mb       *metadata.MetricsBuilder
+	logger   *zap.Logger
+}
+
+// newAerospikeReceiver creates a new aerospikeReceiver connected to the endpoint provided in cfg
+//
+// If the host or port can't be parsed from endpoint, an error is returned.
+func newAerospikeReceiver(params component.ReceiverCreateSettings, cfg *Config, consumer consumer.Metrics) (*aerospikeReceiver, error) {
+	host, portStr, err := net.SplitHostPort(cfg.Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("endpoint: %w", err)
+	}
+
+	port, err := strconv.ParseInt(portStr, 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("port: %w", err)
+	}
+
+	return &aerospikeReceiver{
+		params:   params,
+		config:   cfg,
+		consumer: consumer,
+		host:     host,
+		port:     int(port),
+		mb:       metadata.NewMetricsBuilder(cfg.Metrics),
+		logger:   params.Logger.Named("aerospikereceiver"),
+	}, nil
+}
+
+// scrape scrapes both Node and Namespace metrics from the provided Aerospike node.
+// If CollectClusterMetrics is true, it then scrapes every discovered node
+func (r *aerospikeReceiver) scrape(ctx context.Context) (pmetric.Metrics, error) {
+	errs := &scrapererror.ScrapeErrors{}
+	now := pcommon.NewTimestampFromTime(time.Now().UTC())
+	client, err := newASClient(r.host, r.port, r.config.Timeout)
+	if err != nil {
+		r.logger.Warn(fmt.Sprintf("failed to connect: %s", err.Error()))
+		return r.mb.Emit(), nil
+	}
+	defer client.Close()
+
+	info, err := client.Info()
+	if err != nil {
+		return r.mb.Emit(), err
+	}
+	r.emitNode(info, client, now)
+
+	if r.config.CollectClusterMetrics {
+		r.logger.Sugar().Debugf("Collecting %d peer nodes", len(info.Services))
+		for _, n := range info.Services {
+			r.scrapeDiscoveredNode(n, now, errs)
+		}
+	}
+
+	return r.mb.Emit(), errs.Combine()
+}
+
+// scrapeNode collects metrics from a single Aerospike node
+func (r *aerospikeReceiver) scrapeNode(client Aerospike, now pcommon.Timestamp, errs *scrapererror.ScrapeErrors) {
+	info, err := client.Info()
+	if err != nil {
+		errs.AddPartial(-1, err)
+		return
+	}
+
+	r.emitNode(info, client, now)
+}
+
+// scrapeDiscoveredNode connects to a discovered Aerospike node and scrapes it using that connection
+//
+// If unable to parse the endpoint or connect, that error is logged and we return early
+func (r *aerospikeReceiver) scrapeDiscoveredNode(endpoint string, now pcommon.Timestamp, errs *scrapererror.ScrapeErrors) {
+	host, portStr, err := net.SplitHostPort(endpoint)
+	if err != nil {
+		r.params.Logger.Sugar().Warnf("%w: %s", errBadEndpoint, err)
+		return
+	}
+	port, err := strconv.ParseInt(portStr, 10, 32)
+	if err != nil {
+		r.logger.Sugar().Warnf("%w: %s", errBadPort, err)
+		return
+	}
+	nClient, err := newASClient(host, int(port), r.config.Timeout)
+	if err != nil {
+		r.logger.Warn(err.Error())
+		return
+	}
+	defer nClient.Close()
```

Comment:
For all these error conditions, we should add the error to the passed in ScrapeErrors.